framework/src/utils/KDTree.C

```diff
@@ -0,0 +1,68 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "KDTree.h"
+#include "MooseError.h"
+
+#include "libmesh/nanoflann.hpp"
+
+KDTree::KDTree(std::vector<Point> & master_points)
+  : _kd_tree(nullptr), _point_list_adaptor(master_points)
+{
+}
+
+void
+KDTree::buildTree()
+{
+  // Maximum number of points in each leaf of the Kd tree
+  unsigned int max_leaf_size = 10;
+  if (_kd_tree == nullptr)
+    _kd_tree.reset(new KdTreeT(LIBMESH_DIM,
```

Comment:
Background `libmesh_make_unique` just calls `std::make_unique<T>` but since that's C++14, it's not always there so it falls back to a safe implementation on some compilers.